### DISCUSSION TITLE
NuGet.exe related documentation

### DISCUSSION THREAD

**Discussion by mgnslndh** [cake-build/cake#3271] [Ideas]
Hi!

I was working on Cake Frosting based build and needed to install a NuGet package to a custom location. According to the [documentation](https://cakebuild.net/api/Cake.Common.Tools.NuGet/NuGetAliases/343A93DB#Examples) this was possible, awesome! When I ran the build I found out that Cake could not resolve the path to NuGet.exe, bummer! It seems that the NuGet.exe is not distributed with Visual Studio or any other SDK where Cake might be able to look for it. Thinking about this a bit more I remembered that the scripted Cake build.ps1 used to download NuGet.exe and so this is probably the convention I needed to follow. Now I tried to find some guidance in the documentation about how to best do this but I could not find anything. 

I figured that if I installed NuGet.exe as a tool it would probably work and so it did. 

```
.InstallTool(new Uri("nuget:?package=NuGet.CommandLine&version=5.8.1"))
```

However, I think it would be nice if the documentation pointed this out some how. Either as a section on each NuGet alias page or on the main [NuGet alias page](https://cakebuild.net/dsl/nuget/). Or that Cake make a suggestion to install the NuGet.CommandLine as a tool if it can not be resolved.

I was trying to run a built in tool and was expecting it to just work but it didn't. Then I got confused about how to get the tool even though there is an obvious way to install tools, `InstallTool`. Could be just me that got confused or we could improve the documentation to make it super obvious what you need to do :)

What do you think?

**Comment by augustoproiete** [comment]
@mgnslndh Documenting it (in some fashion) makes a lot of sense to me. It would be good to document both...

... the Frosting way:

```cakescript
.InstallTool(new Uri("nuget:?package=NuGet.CommandLine&version=5.8.1"))
```

... and the Cake scripting way:

```cakescript
#tool "nuget:?package=NuGet.CommandLine&version=5.8.1"
```

Perhaps a quick win would be to update the XML docs in the source-code, to include the above:

![image](https://user-images.githubusercontent.com/177608/110888549-c8428100-82c2-11eb-9025-ed893682d764.png)

Something similar to how @nils-a is doing in the [Cake.7zip examples](https://cakebuild.net/extensions/cake-7zip/).

![image](https://user-images.githubusercontent.com/177608/110888760-3c7d2480-82c3-11eb-8a1b-7511db72ac31.png)

One thing to consider is that we might want to add to use placeholders for the version numbers e.g. `nuget:?package=NuGet.CommandLine&version=x.y.z` instead of hard-code a specific version (e.g. `5.8.1`). Related: https://github.com/cake-build/website/issues/587

---

Adding an alert to the [main NuGet aliases page](https://cakebuild.net/dsl/nuget/) also seems like a good idea - though might be a little complicated given it's auto-generated.

---

One other thing we could consider is to add a little exception handling in the NuGet aliases, and display a more friendly message to the user when we cannot find the NuGet executable.

---

/cc @pascalberger

**Reply by pascalberger** [reply]
> I was trying to run a built in tool and was expecting it to just work but it didn't. 

The above sentence is not true as Cake is just a build automation tool and you never run a tool shipped by Cake, but only call a built in alias to run a tool. While we could document this and already do for most aliases (see [here](https://cakebuild.net/dsl/nunit/)), it is important for users to understand the difference between providing a wrapper which calls a tool and shipping a tool. There is no difference here between calling a NuGet alias as calling a MsBuild alias or any other alias.

While I agree we should mention this on the [reference page](https://cakebuild.net/dsl/nuget/), like we do for other aliases, I wouldn't put more effort into this specific use case (like documenting every alias, implementing specific error message), as this would only help in this one specific case, but not for all other addins.
Instead it would probably be more helpful to spent the effort by adding more content explaining the overall concepts (Cake just calls tools which need to be installed, and also stuff like that if the tool doesn't run outside of Cake it won't magically run when called by Cake, which is an often support case). This could for example be done in https://cakebuild.net/docs/fundamentals/.

**Reply by augustoproiete** [reply]
> The above sentence is not true as Cake is just a build automation tool and you never run a tool shipped by Cake...

No question about that, but the reality is that when someone is starting out with Cake, they likely only have Visual Studio installed, which means that:

- `DotNetCore***` aliases work 99% of the time, because `dotnet` is available
- `MSBuild***` aliases work 99% of the time, because `msbuild` is available
- `NuGet***` aliases **fail** 99% of the time, because `nuget` is **rarely** available

Having a small explanation at the top of the alias class level would be helpful/effective, and low-cost as @nils-a demonstrates here: https://github.com/cake-contrib/Cake.7zip/blob/b7b39cc0b2fd6f66c17c6cd868a1df0b38e4b5b6/src/Cake.7zip/SevenZipAliases.cs#L10-L32

**Reply by augustoproiete** [reply]
@mgnslndh Would you be interested in sending a PR that documents the needs for the NuGet alias in the [NuGet aliases](https://cakebuild.net/dsl/nuget/) page the same way that is documented in the [NUnit aliases](https://cakebuild.net/dsl/nunit/) as @pascalberger pointed out?

[The XML docs for NUnit are here](https://github.com/cake-build/cake/blob/8bc15841e97944291946a89f335562244965247e/src/Cake.Common/Tools/NUnit/NUnitAliases.cs#L15-L25), and you can see examples for other aliases using [this search](https://github.com/cake-build/cake/search?q=to+use+the+commands+for+this+alias) on the GitHub repo.
